flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncCollectorBuffer.java

```diff
@@ -0,0 +1,453 @@
+	 * corresponding to the input StreamRecord.
+	 * <p>
+	 * If buffer is full, caller will wait until a new space is available.
+	 *
+	 * @param record StreamRecord
+	 * @return An AsyncCollector
+	 * @throws Exception InterruptedException or IOException from AsyncCollector.
+	 */
+	public AsyncCollector<IN, OUT> addStreamRecord(StreamRecord<IN> record) throws InterruptedException, IOException {
+		while (queue.size() >= bufferSize) {
+			// hold the input StreamRecord until it is placed in the buffer
+			extraStreamElement = record;
+
+			lock.wait();
+		}
+
+		if (error != null) {
+			throw error;
+		}
+
+		AsyncCollector<IN, OUT> collector = new AsyncCollector(this);
+
+		queue.put(collector, record);
+
+		extraStreamElement = null;
+
+		return collector;
+	}
+
+	/**
+	 * Add a {@link Watermark} into queue. A new AsyncCollector will be created and returned.
+	 * <p>
+	 * If queue is full, caller will wait here.
+	 *
+	 * @param watermark Watermark
+	 * @return AsyncCollector
+	 * @throws Exception InterruptedException or IOException from AsyncCollector.
+	 */
+	public AsyncCollector<IN, OUT> addWatermark(Watermark watermark) throws InterruptedException, IOException {
+		return processMark(watermark);
+	}
+
+	/**
+	 * Add a {@link LatencyMarker} into queue. A new AsyncCollector will be created and returned.
+	 * <p>
+	 * If queue is full, caller will wait here.
+	 *
+	 * @param latencyMarker LatencyMarker
+	 * @return AsyncCollector
+	 * @throws Exception InterruptedException or IOException from AsyncCollector.
+	 */
+	public AsyncCollector<IN, OUT> addLatencyMarker(LatencyMarker latencyMarker) throws InterruptedException, IOException {
+		return processMark(latencyMarker);
+	}
+
+	private AsyncCollector<IN, OUT> processMark(StreamElement mark) throws InterruptedException, IOException {
+		while (queue.size() >= bufferSize) {
+			// hold the input StreamRecord until it is placed in the buffer
+			extraStreamElement = mark;
+
+			lock.wait();
+		}
+
+		if (error != null) {
+			throw error;
+		}
+
+		AsyncCollector<IN, OUT> collector = new AsyncCollector(this, true);
+
+		queue.put(collector, mark);
+
+		extraStreamElement = null;
+
+		return collector;
+	}
+
+	/**
+	 * Notify the emitter thread and main thread that an AsyncCollector has completed.
+	 *
+	 * @param collector Completed AsyncCollector
+	 */
+	void markCollectorCompleted(AsyncCollector<IN, OUT> collector) {
+		synchronized (lock) {
+			collector.markDone();
+
+			// notify main thread to keep working
+			lock.notifyAll();
+		}
+	}
+
+	/**
+	 * Caller will wait here if buffer is not empty, meaning that not all async i/o tasks have returned yet.
+	 *
+	 * @throws Exception InterruptedException or IOException from AsyncCollector.
+	 */
+	void waitEmpty() throws InterruptedException, IOException {
+		while (queue.size() != 0) {
+			if (error != null) {
+				throw error;
+			}
+
+			lock.wait();
+		}
+	}
+
+	public void startEmitterThread() {
+		this.emitThread.start();
```

Comment:
You mean starting the `emitter thread` while constructing `AsyncCollectorBuffer`? Since we have to get a `AsyncCollectorBuffer` ready to put inputs from previous state while `open()` the operator. After all inputs are in the buffer, `emitter thread` then can start working.